Develop initial working plugin

mariadanieldeepak opened the pull request:
A working version of the plugin has been developed

mariadanieldeepak committed:
Include initial search-categories-metabox.php

mariadanieldeepak committed:
Include initial class-search-categories-metabox.php

mariadanieldeepak committed:
Develop basic working plugin

sudar commented on assets/css/admin-style.css:
```diff
@@ -0,0 +1,19 @@
+.scm-categories-panel {
+    /*min-height: 42px;
```
Commented code :)

mariadanieldeepak replied:
Removed the commented code.

sudar commented on assets/js/admin-script.js:
```diff
@@ -0,0 +1,155 @@
+/* global selectedCategories:true */
```
It would be nice to have the assets named something with the plugin name instead of generic name like `admin-script`.

mariadanieldeepak replied:
Renamed both JS and CSS files.

sudar commented on assets/js/admin-script.js:
```diff
@@ -0,0 +1,155 @@
+                otherCategories    = [];
+
+            $( 'ul#scm-categories-ul li' ).each(function () {
+                checkBoxState = $( this ).find( 'input' ).prop( 'checked' );
```
`checkBoxState` is not defined.

mariadanieldeepak replied:
Defined the variable `checkBoxState`

sudar commented on assets/js/admin-script.js:
```diff
@@ -0,0 +1,155 @@
+             * When user deselects, remove the item from
+             * global selectedCategories
+             */
+            if ( checkBoxState == false ) {
```
Either use Yoda or just check `( ! checkBoxState )`

mariadanieldeepak replied:
Fixed by using Yoda.

sudar commented on assets/js/admin-script.js:
```diff
@@ -0,0 +1,155 @@
+                    $( 'ul#scm-categories-ul' ).html( jsonResult.category_list );
+                } else {
+                    //$('ul#scm-categories-ul').html('No results found.');
+                    $( 'ul#scm-categories-ul' ).html( '<li>No results found.</li>' );
```
This text is not translated. Ideally it would be good to have this translated.

mariadanieldeepak replied:
Replaced with Translatable strings.

sudar commented on assets/js/admin-script.js:
```diff
@@ -0,0 +1,155 @@
+                var jsonResult = data;
+
+                if( '' != jsonResult.category_list ) {
+                    $( 'ul#scm-categories-ul' ).html( jsonResult.category_list );
```
This might have some security implications.

mariadanieldeepak replied:
Thanks for bringing this out. Do you mean to use `wp_send_json_success( $response )`? Will that help to avoid this security issue?

sudar commented on assets/js/admin-script.js:
```diff
@@ -0,0 +1,155 @@
+                sortCategoriesListBySelectedItems();
+
+            }).error(function() {
+                var errorMessage = '<li>Something went wrong.</li>';
```
Again would be nice to have this text translatable.

mariadanieldeepak replied:
Replaced with Translatable strings.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,326 @@
+     * @param $post_id
+     * @return array
+     */
+    function _get_post_taxonomy( $post_id ) {
```
It's a matter of personal preference, but I would prefer to explicitly state access modifiers like `public` or `private`, instead of using conventions like `_`.

mariadanieldeepak replied:
Oops, somehow I missed to change this during development. Fixed by adding access modifier.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,326 @@
+     * @return array
+     */
+    function _get_post_taxonomy( $post_id ) {
+        $taxonomy = get_post_taxonomies( $post_id );
```
It would be more readable if you called this variable `$taxonomies` and then just saw `return $taxonomies[0]`

mariadanieldeepak replied:
Awesome. Modified as suggested.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,326 @@
+
+        $post_taxonomy       = $this->_get_post_taxonomy( $post->ID );
+        $selected_categories = $_POST['post-category'];
+        if ( isset( $selected_categories ) && ! empty( $selected_categories ) ) {
```
`isset` should be called on `_POST`

mariadanieldeepak replied:
Fixed by adding the isset check.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,326 @@
+            ?>
+                <li id='<?php echo $category->term_id; ?>'>
+                    <label class="selecttit">
+                        <input id='in-category-<?php echo $category->term_id; ?>' class="scm-category-list-item" value=<?php echo $category->term_id; ?> name="post-category[]" type="checkbox" <?php if ( $is_post_in_category ) { echo 'checked="checked"'; } ?>> <?php echo $category->name; ?>
```
You may have to escape some of these text before printing them.

mariadanieldeepak replied:
Escaped the Category name.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,326 @@
+            ?>
+                <li id='<?php echo $category->term_id; ?>'>
+                    <label class="selecttit">
+                        <input id='in-category-<?php echo $category->term_id; ?>' class="scm-category-list-item" value=<?php echo $category->term_id; ?> name="post-category[]" type="checkbox" <?php if ( $is_post_in_category ) { echo 'checked="checked"'; } ?>> <?php echo $category->name; ?>
```
You may have to escape some of these text before printing them.

mariadanieldeepak replied:
Escaped the Category name.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,326 @@
+
+
+        <!--
+        <div id="category-adder" class="">
```
Commented code ;)

mariadanieldeepak replied:
Removed :)

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,326 @@
+
+        check_ajax_referer( 'search-categories-metabox', 'security' );
+
+        $search_query        = $_POST['query'];
```
User input not sanitized before using it.

mariadanieldeepak replied:
Sanitized the input using `sanitize_text_field()`

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,326 @@
+
+            <h1><?php echo esc_html( get_admin_page_title() ); ?></h1>
+
+            <form action='<?php echo esc_html( $_SERVER['REQUEST_URI'] ); ?>' method='POST'>
```
You should use `esc_attr` as the escaping function

mariadanieldeepak replied:
This method is used to add Admin Menu. I'm planning to bring this option in the next release. Remove this method for now.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,326 @@
+    public function add_admin_menu() {
+
+        add_options_page(
+            __( 'Search Categories Metabox', Search_Categories_Metabox::TEXT_DOMAIN ),
```
Text domain is parsed by static code tools. So it should always be hardcoded as a string and not stored as a variable. More details at https://markjaquith.wordpress.com/2011/10/06/translating-wordpress-plugins-and-themes-dont-get-clever/

mariadanieldeepak replied:
:+1:

sudar commented on search-categories-metabox.php:
```diff
@@ -0,0 +1,33 @@
+<?php
+/*
```
WordPress coding standard recommends using `/**` instead of `/*`

mariadanieldeepak replied:
Did WP recently modify to use /* ?

Please take a look at this [Codex page](https://developer.wordpress.org/plugins/the-basics/header-requirements/).

mariadanieldeepak committed:
Remove comment lines from CSS

mariadanieldeepak committed:
Convert Spaces to Tabs in CSS

mariadanieldeepak committed:
Rename JS file to search-categories-metabox.js

mariadanieldeepak committed:
Define var `checkboxState` before using in JS

mariadanieldeepak committed:
Use Yoda conditions to check equality

mariadanieldeepak committed:
Use translatable string in JS error messages

mariadanieldeepak committed:
Specify access modifiers implicitly

mariadanieldeepak committed:
Improve readability by renaming variables

mariadanieldeepak committed:
Implement isset() check on $_POST vars

mariadanieldeepak committed:
Rename function name for readability

mariadanieldeepak committed:
Escape data before outputting to User

mariadanieldeepak committed:
Remove commented HTML

mariadanieldeepak committed:
Sanitize User Input

mariadanieldeepak committed:
Replace esc_html with esc_attr in Form Action

mariadanieldeepak committed:
Replace dynamic Text domain with static value

Text domain is parsed by static code tools. So it should always be hardcoded as a string and not stored as a variable. More details at https://markjaquith.wordpress.com/2011/10/06/translating-wordpress-plugins-and-themes-dont-get-clever/

mariadanieldeepak committed:
Fix data retrieval from JSON result

mariadanieldeepak committed:
Fix selectedCategories returns empty from AJAX

mariadanieldeepak committed:
Remove Admin Menu

mariadanieldeepak committed:
Convert Spaces to Tabs

mariadanieldeepak committed:
Improve readability of JS file

sudar commented on assets/js/search-categories-metabox.js:
```diff
@@ -0,0 +1,186 @@
+		 * @return array List of User selected Categories.
+		 */
+		function getSelectedCategories() {
+			// Category ID's.
```
It is not a comment, but just a suggestion. Also it is just my personal preference/opinion.

I feel that you are over commenting the code. Generally comments should just say why and not what or how. What or how should be clear enough if the code id readable.

The problem with over commenting is that, it is very easy for it to get outdated when you make code changes. Also for someone who can read code it is redundant.

Eg:

```
// Return the array of selected Category ID's.
return categories;
```

It is very obvious what you are returning.

mariadanieldeepak replied:
Got it, thank you :)

sudar commented on assets/js/search-categories-metabox.js:
```diff
@@ -0,0 +1,186 @@
+				errorMessage = 'Something went wrong.';
+			}
+
+			$( 'ul#scm-categories-ul' ).html( '<li>' + errorMessage  + '</li>' );
```
This could result in a security issue if someone hijacks your AJAX request.

Eg: What happens if they change the message to a script tag?

Ideally you should just use `text` method or create the element and then append them.

mariadanieldeepak replied:
I wasn't aware that this is a security issue. I took time to read about it and got some idea. I fixed it by creating a DOM `<li>` element and then I used `text` to write content to it and appended it where ever required.

sudar commented on assets/js/search-categories-metabox.js:
```diff
@@ -0,0 +1,186 @@
+						$( 'ul#scm-categories-ul' ).html( jsonResult.category_list );
+					} else {
+						// Error message to display when no results are found.
+						var noResultsFound;
```
In JavaScript it is always recommended to declare all variables at the top of the function.

mariadanieldeepak replied:
Got it, thanks!

sudar commented on assets/js/search-categories-metabox.js:
```diff
@@ -0,0 +1,186 @@
+						if ( $.type( ajax_object.no_results_error_message ) != 'undefined' ) {
+							noResultsFound = ajax_object.no_results_error_message;
+						} else {
+							noResultsFound = 'No results found.';
```
This string is not available for translation.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+	 * @return array
+	 */
+	private function get_post_taxonomy( $post_id ) {
+		$taxonomies = get_post_taxonomies( $post_id );
```
`get_post_taxonomies` can return more than one result and the order in which they are sorted is not guaranteed. Then how are you picking up just the first element in the array?

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+	 * Gets the array of categories.
+	 *
+	 * @since 0.1.0
+	 * @access private
```
PHPDoc says it is `private`, but in reality it is `protected`

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+	 * @return array
+	 */
+	protected function get_taxonomy_categories( $search_term = '' ) {
+		$categories = array();
```
Why declare these variables and then reassign a value when they are used for the first time?

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+			$args = wp_parse_args( $additional_args, $args );
+		}
+
+		$categories = get_categories( $args );
```
You can directly return it instead of storing it in a variable and then returning it.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+
+		<p>
+			<input id="scm-search-categories-search" type="text" name="scm-search-categories-search" size="16" />
+			<input type="button" class="button" id="scm-search-clear" value="Clear" />
```
The string `Clear` is not available for translation.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+			?>
+				<li id='<?php echo $category->term_id; ?>'>
+					<label class="selecttit">
+						<input id='in-category-<?php echo $category->term_id; ?>' class="scm-category-list-item" value=<?php echo $category->term_id; ?> name="post-category[]" type="checkbox" <?php if ( $is_post_in_category ) { echo 'checked="checked"'; } ?>> <?php echo esc_html( $category->name ); ?>
```
Instead of this condition `if ( $is_post_in_category ) { echo 'checked="checked"'; }` you can use the `checked` function

mariadanieldeepak replied:
Could you please brief on this? I didn't get it exactly.

sudar replied:
There is a function is WordPress called `checked`. Check the example listed at https://developer.wordpress.org/reference/functions/checked/

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+	 */
+	public function add_search_categories_metabox() {
+
+		add_meta_box( 'searchcategorydiv', 'Categories', array( $this, 'search_category_div_callback' ), 'post', 'side', 'default', null );
```
String not available for translation

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+		$search_query        = isset( $_POST['query'] ) ? sanitize_text_field( $_POST['query'] ) : '';
+		$selected_categories = ( isset( $_POST['categories'] ) && ! empty( $_POST['categories'] ) ) ? $_POST['categories'] : array();
+		$result              = [];
+		$post_id             = $_POST['post_id'];
```
Sanitize user input

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+			$result['category_list'] = ob_get_clean();
+		}
+
+		//echo json_encode( 'Daniel' );
```
Commented code.

sudar commented on includes/class-search-categories-metabox.php:
```diff
@@ -0,0 +1,281 @@
+
+			<form action='<?php echo esc_attr( admin_url( 'options-general.php?page=' ) . self::PLUGIN_SLUG ); ?>' method='POST'>
+				<?php
+				submit_button( 'Save Settings' );
```
String not available for translation.

sudar commented on search-categories-metabox.php:
```diff
@@ -0,0 +1,33 @@
+
+}
+
+run_search_categories_metabox();
```
Call this function in `plugins_loaded` hook.

sudar commented on search-categories-metabox.php:
```diff
@@ -0,0 +1,33 @@
+<?php
+/*
+Plugin Name: Search Categories Metabox
+Plugin URI:  https://developer.wordpress.org/plugins/the-basics/
```
If possible change the url to something different.

sudar reviewed (changes requested):
Left some more review comments.

mariadanieldeepak committed:
Prevent AJAX hijacking in JS

mariadanieldeepak committed:
Display selected categories on Post Editor

mariadanieldeepak committed:
Fix GitHub review comments
